src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -1309,6 +1317,15 @@ namespace WaterThermalTanks {
                             ShowContinueError("...occurs in " + cCurrentModuleObject + '=' + WaterHeaterDesuperheater(DesuperheaterNum).Name);
                             ErrorsFound = true;
                         }
+                    } else if (UtilityRoutines::SameString(cAlphaArgs(9), "Coil:Cooling:WaterToAirHeatPump:EquationFit")) {
+                        WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceType = cAlphaArgs(9);
+                        WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceName = cAlphaArgs(10);
+                        errFlag = false;
+                        WAHPHeatingSourceNum = WaterToAirHeatPumpSimple::GetCoilIndex(cAlphaArgs(9), cAlphaArgs(10), errFlag);
```

Comment:
Here instead of `WAHPHeatingSourceNum =`, it should be `WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSourceIndexNum = . . . GetCoilIndex`. Likewise for all of the other source types in this section. 
Plus, the error check can be done once after all the ifs and then if you also add `ValidSourceType(DesuperheaterNum) = true;` here, then the entire section from lines 1354:1453 can be deleted.

For example (starting at line 1298):
````
                    //       get heat reclaim object
                    errFlag = false;
                    WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceType = cAlphaArgs(9);
                    WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceName = cAlphaArgs(10);
                    if (UtilityRoutines::SameString(cAlphaArgs(9), "Coil:Cooling:DX:SingleSpeed") ||
                        UtilityRoutines::SameString(cAlphaArgs(9), "Coil:Cooling:DX:TwoSpeed") ||
                        UtilityRoutines::SameString(cAlphaArgs(9), "Coil:Cooling:DX:TwoStageWithHumidityControlMode")) {
                        //         load DX coil structure for connection to desuperheater heating coil (refrigerated rack have been loaded)
                        GetDXCoilIndex(WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceName, WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSourceIndexNum, errFlag, cCurrentModuleObject);
                    } else if (UtilityRoutines::SameString(cAlphaArgs(9), "Coil:Cooling:DX:VariableSpeed")) {
                        WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSourceIndexNum = VariableSpeedCoils::GetCoilIndexVariableSpeed(cAlphaArgs(9), cAlphaArgs(10), errFlag);
                    } <and so on for the other coil types

                    } else {
                        ShowSevereError(cCurrentModuleObject + " = " + WaterHeaterDesuperheater(DesuperheaterNum).Name + ':');
                       < . . . >
                   } 

                    if (errFlag) {
                        ShowContinueError("...occurs in " + cCurrentModuleObject + '=' + WaterHeaterDesuperheater(DesuperheaterNum).Name);
                        ErrorsFound = true;
                    } else {
                       ValidSourceType(DesuperheaterNum) = true;
                    }
````
This will be much cleaner and will eliminate several of the int variables declared at the top. If this isn't clear, I'd be happy to make these code changes to help make this function more clear.

Reply:
Hi, Mike, after I checked failed CI, I surprisingly found these two parts are not doing the same thing. The first part gets the coil index in coil lists. The second one is attaching the coil sources to corresponding desuperheater objects. But the confusing thing is that I didn't see any point doing the first search except for coil validation check.